streams/delimiter_stream.ts

```diff
@@ -83,32 +80,109 @@ export class DelimiterStream extends TransformStream<Uint8Array, Uint8Array> {
     chunk: Uint8Array,
     controller: TransformStreamDefaultController<Uint8Array>,
   ) {
-    this.#bufs.add(chunk);
-    let localIndex = 0;
-    while (this.#inspectIndex < this.#bufs.size()) {
-      if (chunk[localIndex] === this.#delimiter[this.#matchIndex]) {
-        this.#inspectIndex++;
-        localIndex++;
-        this.#matchIndex++;
-        if (this.#matchIndex === this.#delimLen) {
+    const bufs = this.#bufs;
+    const length = chunk.byteLength;
+    const disposition = this.#disp;
+    const delimiter = this.#delimiter;
+    const delimLen = delimiter.length;
+    const lps = this.#delimLPS;
+    let chunkStart = 0;
+    let matchIndex = this.#matchIndex;
+    let inspectIndex = 0;
+    while (inspectIndex < length) {
+      if (chunk[inspectIndex] === delimiter[matchIndex]) {
+        // Next byte matched our next delimiter byte
+        inspectIndex++;
+        matchIndex++;
+        if (matchIndex === delimLen) {
           // Full match
-          const start = this.#inspectIndex - this.#delimLen;
-          const end = this.#disp === "suffix" ? this.#inspectIndex : start;
-          const copy = this.#bufs.slice(0, end);
-          controller.enqueue(copy);
-          const shift = this.#disp === "prefix" ? start : this.#inspectIndex;
-          this.#bufs.shift(shift);
-          this.#inspectIndex = this.#disp === "prefix" ? this.#delimLen : 0;
-          this.#matchIndex = 0;
+          matchIndex = 0;
+          const delimiterStartIndex = inspectIndex - delimLen;
+          const delimitedChunkEnd = disposition === "suffix"
+            ? inspectIndex
+            : delimiterStartIndex;
+          if (delimitedChunkEnd <= 0 && bufs.length === 0) {
```

Comment:
**comment:** Kind of an obvious downside here: I've written out all the possible cases in excruciating detail to optimise each of them individually, and added a fair bit of comments. The number of lines has gone way up.